### DISCUSSION TITLE
Allowing for provider level customizations

### DISCUSSION THREAD

**Discussion by sixlive** [prism-php/prism#36] [Ideas]
Hey folks! I'd love your thoughts on this.

I'm working on allowing provider-level customizations, starting with HTTP client options. Here's what I'm considering:

```php
$response = Prism::text()
  ->using(Provider::Ollama, 'qwen2.5:14b')
  ->withProviderCustomizations(function (Ollama $provider) {
      $provider->withClientOptions(['timeout' => '10']);
  })
  ->withPrompt('Who are you?')
  ->generate();
```

Specifically speaking about client options, we could also just make client customizations its own method:

```php
$response = Prism::text()
  ->using(Provider::Ollama, 'qwen2.5:14b')
  ->withClientOptions(['timeout' => '10'])
  ->withPrompt('Who are you?')
  ->generate();
```

This approach could also be used to expose provider-specific options and settings in the future.

What do you think? Any suggestions or concerns? Thanks in advance for your feedback!

**Comment by mattglover11** [comment]
@sixlive I'm wondering out loud here - do we need to instantiate/inject another instance of the provider, or can we do this under the covers? I'm replying via the phone, so treat this more like pseudo thought :) Inside the box, presuming Prism needs to instantiate the provider, this was what I thought might possibly work:

$response = Prism::text()
    ->using(Provider::Ollama, 'qwen2.5:14b')
    ->withClientOptions(['timeout' => 10])
    ->withProviderOptions([
        'temperature' => 0.7,
        'max_tokens' => 100
    ])
    ->withPrompt('Who are you?')
    ->generate();

**Reply by slashequip** [reply]
I like this approach too, being more explicit with the methods as @mattglover11 shared. I think once you get past this level you are likely into custom provider territory anyway? Being able to config the client and base provider options that are sent should be more than enough for most.

We use a rule at Hospitable called "Alienate the 3%" which goes along the lines of "it's not possible to be everything to everyone, and sometimes it's necessary to prioritize the needs of the majority (97%) over a small minority (3%) of users". I think that would apply here.

**Reply by sixlive** [reply]
Thank you both! Super valuable feedback. I'm going to take the explicit approach, it aligns well with the way the rest of the package works.

**Comment by sixlive** [comment]
Another option that I was thinking of was passing client/provider options as part of the using method.

```php
Prism::text()
  ->using(
    Provider::Ollama,
    'qwen2.5',
    providerOptions: [
      'client' => [],
    ],
  )
  ->generate()
```

**Reply by sixlive** [reply]
@slashequip @mattglover11 what do you think of this? It's the most straight forward approach, in my opinion.

The builder handles most common provider options like max tokens, temp, topP, etc.

**Reply by slashequip** [reply]
Feels like extra fluff to explain vs keeping them in more specific methods. I personally like more specific methods vs more args. IMO though.

**Reply by sixlive** [reply]
Yeah, while I like this approach because its straight forward and easy to implement, I don't necessarily like 'magic keys'. You're totally right, its not self documenting like a method is.

**Comment by sixlive** [comment]
I ended up merging https://github.com/echolabsdev/prism/pull/37 with a new method `->withClientOptions()`. Punting on provider specific configuration since I don't have a need specific need for that at the moment.
